### PR TITLE
Fix rapid7/metasploit-framework#16493, remove RWX section and apply section Characteristics

### PR DESCRIPTION
**Pull request by timwr**
This is a step in the right direction for fixing https://github.com/rapid7/metasploit-framework/issues/16493
This removes *one* of the two RWX regions created when the ReflectiveLoader loads a dll from memory.
The other one (I think is this: https://github.com/rapid7/metasploit-framework/blob/f737e78b876711f1a8d05809fd4632bc95181e65/lib/msf/core/payload/windows/reflective_pe_loader.rb#L32) we can remove later (I suspect a small section needs to be RX, and the rest can be R or RW), additionally we may be able to deallocate this entire block afterwards.

## Verification steps

1) Build meterpreter with these changes
2) Get a meterpreter session (I tested windows/x64/meterpreter/reverse_tcp on Windows 10).
3) View the meterpreter process with process hacker, and look at the Memory tab. Notice one of the RWX allocation is now RW, with one section allocated as RX

## Before

VirtualBox_Windows 10_25_04_2022_04_28_01

## After
VirtualBox_Windows 10_25_04_2022_08_57_24

### CODE REVIEW

**Comment by timwr** [comment]
I'm un-marking this a draft. This should be only change required to remove RWX regions from the ReflectiveDLLInjection repository. There are other instances of RWX memory in this repository:
```
./inject/src/LoadLibraryR.c:                    if( VirtualProtect( lpBuffer, dwLength, PAGE_EXECUTE_READWRITE, &dwOldProtect1 ) )
./inject/src/LoadLibraryR.c:                    if (!VirtualProtectEx(hProcess, lpRemoteLibraryBuffer, dwLength, PAGE_EXECUTE_READ, NULL))
./dll/src/ReflectiveLoader.c:   uiBaseAddress = (ULONG_PTR)pVirtualAlloc( NULL, ((PIMAGE_NT_HEADERS)uiHeaderValue)->OptionalHeader.SizeOfImage, MEM_RESERVE|MEM_COMMIT, PAGE_EXECUTE_READWRITE );
```
but these are part of the inject project, and are not used during meterpreter staging.

**Comment by smcintyre-r7** [comment]
TLDR; I'm going to land this as is, thanks @timwr !

# Testing
I built the Meterpreter bins for x86 and x64. I compared them on Windows 11 using process hacker as seen in the screenshot above to prove that there was a change in RWX allocations.

Next I tested x86 and x64 Meterpreters for both the reverse_tcp and reverse_http payloads on Windows 7 as delivered using psexec. For all of these cases I tested migrating into a new process. Someone had mentioned the config block being RWX and that these changes may have affected that. The config block is used during the migration process. I didn't identify any issues here which makes sense because on closer inspection, the Meterepter [handles the allocations](https://github.com/rapid7/metasploit-payloads/blob/5dca4f5989cc9ce1c80286e48508d34568af78c9/c/meterpreter/source/metsrv/base_dispatch.c#L634) for the config block itself. In addition to migrate, I also ran the following post test modules for all sessions and transports I mentioned:

  * post/test/meterpreter
  * post/test/railgun
  * post/test/file
  * post/test/registry
  * post/test/services
  * post/test/cmd_exec
  * post/test/extapi

A couple of the tests failed, but the Meterpreter session did not crash. Tests are known to be unstable but if there were a memory permission related issue, I'd have expected the session to crash.

So with these changes in place, the next step will be to compile them into Meterpreter and ship those bins.

It's possible that these change may break some other RDLLs particularly some LPE exploits we ship as RDLLs. The chance seems unlikely and those bins were built with a previous version of this code. I don't think it's worth while to recompile all of those with this change. RDLL LPEs built moving forward should incorporate it and should be tested during development.
